cpp/include/raft/mdarray.hpp

```diff
@@ -0,0 +1,425 @@
+  {
+    return map_.stride(r);
+  }
+
+  [[nodiscard]] MDSPAN_INLINE_FUNCTION static constexpr auto is_always_unique() noexcept -> bool
+  {
+    return mapping_type::is_always_unique();
+  }
+  [[nodiscard]] MDSPAN_INLINE_FUNCTION static constexpr auto is_always_contiguous() noexcept -> bool
+  {
+    return mapping_type::is_always_contiguous();
+  }
+  [[nodiscard]] MDSPAN_INLINE_FUNCTION static constexpr auto is_always_strided() noexcept -> bool
+  {
+    return mapping_type::is_always_strided();
+  }
+
+ private:
+  template <typename, typename, typename, typename>
+  friend class mdarray;
+
+ private:
+  container_policy_type cp_;
+  mapping_type map_;
+  container_type c_;
+};
+
+/**
+ * @brief mdarray with host container policy
+ */
+template <class ElementType,
+          class Extents,
+          class LayoutPolicy    = detail::stdex::layout_right,
+          class ContainerPolicy = detail::host_vector_policy<ElementType>>
+using host_mdarray =
+  mdarray<ElementType, Extents, LayoutPolicy, detail::host_accessor<ContainerPolicy>>;
+
+/**
+ * @brief mdarray with device container policy
+ */
+template <class ElementType,
+          class Extents,
+          class LayoutPolicy    = detail::stdex::layout_right,
+          class ContainerPolicy = detail::device_uvector_policy<ElementType>>
+using device_mdarray =
+  mdarray<ElementType, Extents, LayoutPolicy, detail::device_accessor<ContainerPolicy>>;
+
+/**
+ * @brief Shorthand for 1-dim host mdarray.
+ */
+template <typename ElementType>
+using host_vector = host_mdarray<ElementType, detail::vector_extent>;
+
+/**
+ * @brief Shorthand for 1-dim device mdarray.
+ */
+template <typename ElementType>
+using device_vector = device_mdarray<ElementType, detail::vector_extent>;
+
+/**
+ * @brief Shorthand for c-contiguous host matrix.
+ */
+template <typename ElementType>
+using host_matrix = host_mdarray<ElementType, detail::matrix_extent>;
+/**
+ * @brief Shorthand for c-contiguous device matrix.
+ */
+template <typename ElementType>
+using device_matrix = device_mdarray<ElementType, detail::matrix_extent>;
+
+template <typename ElementType>
+using host_vector_view = host_mdspan<ElementType, detail::vector_extent>;
+
+template <typename ElementType>
+using device_vector_view = device_mdspan<ElementType, detail::vector_extent>;
+
+/**
+ * @brief Shorthand for c-contiguous host matrix view.
+ */
+template <typename ElementType>
+using host_matrix_view = host_mdspan<ElementType, detail::matrix_extent>;
+/**
+ * @brief Shorthand for c-contiguous device matrix view.
+ */
+template <typename ElementType>
+using device_matrix_view = device_mdspan<ElementType, detail::matrix_extent>;
+
+/**
+ * @brief Create a 2-dim c-contiguous mdspan instance for host pointer.
+ */
+template <typename ElementType>
+auto make_host_matrix_view(ElementType* ptr, size_t n_rows, size_t n_cols)
+{
+  detail::matrix_extent extents{n_rows, n_cols};
+  return host_matrix_view<ElementType>{ptr, extents};
+}
+/**
```

Comment:
Added the optional layout parameter along with a test showing how to see whether input is c/f-contiguous.  Hope that helps.

Also, due to the implementation of submdspan, the returned slice has `default_accessor` instead of our host/device tagged accessor.  I will investigate it further, for now, it will be a minor inconvenience.

Reply:
@cjnolet There are additional details in here.

> I might suggest we expose a version of the layout to the public API so that users don't have to call into the detail namespace

Agreed, I will expose some more things from mdspan ref impl.  This will replace `stdex` with `raft`. (no `detail` in here so I'm not sure if I understand you correctly).

> If we can, it would also be nice to rename the public version to "layout::colmajor" and "layout::rowmajor" 

This might be misleading. (correct me if I'm wrong)  According to the design of `stdex::mdspan`, `stdex::layout_right` is c-contiguous, which implies row-major and contiguous.  An array can be row-major but not contiguous.  From this perspective, I think it might be better if we stick to the mdspan naming.

Reply:
I saw your PR on using it for distance metrics.  I think it would be better if we can pass in mdspan all the way down to the kernels and be careful around the `extent` and `stride` when using it with cublas.  Converting a span directly to a pointer is dangerous.

Reply:
> (no detail in here so I'm not sure if I understand you correctly).

I'm referring to `detail::stdex::layout_right` here. 

> This might be misleading. 

I'm proposing both to be used. Many developers are going to come into this API understanding things like row-/col-major and c/f-contiguity but not understanding the right vs left verbiage. We can cover both cases if we provide both here. We can constrain the API if needed, but if we plan to support non-contiguous layouts in the future, I agree the names could best reflect that. 

>  I think it would be better if we can pass the mdspan all the way down to the kernels and be careful around the extent and stride when using it with cublas. 

Thanks for taking a look at the PR! It was intended to initiate discussions like this. RAFT's primitives are pretty extensive at this point and re-writing all of the kernels up front to suit this new API would be an extremely massive undertaking. I understand the mdspan's underlying indexing can be very flexible but I'd like to take a little step back and think through ways we can provide this functionality incrementally (within RAFT) without rewriting the entire library. With contiguity assumed (as it is currently), the row- vs col-major cases seem like a pretty simple introduction to using the mdspan as a facade. Shouldn't we be able to constrain these cases by testing the layout? I think as we expand the kernels to use the mdspan directly for indexing then the API will just become more flexible. For now, the entire RAFT is assuming pointers w/ contiguous layouts already.

Reply:
Added type alias along with brief documents.